File: src/component/sprite/Sprite.java
package component.sprite;

import org.newdawn.slick.Graphics;
import org.newdawn.slick.Image;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.gui.GUIContext;

import component.Component;

/**
 * {@code Sprite} inherits from {@code Component} to extend features
 * to represent a sprite with an left and right animation.
 */
public class Sprite extends Component {
	private Image image;
	
	public Sprite(GUIContext context, int width, int height) {
		super(context, width, height);
	}
	
	public Sprite(GUIContext context, int width, Image image) {
		this(context, width, width * image.getHeight() / image.getWidth());
		
		this.image = image;
		
		image.setCenterOfRotation(getWidth() / 2, getHeight() / 2);
	}
	public Sprite(GUIContext context, int width, int height, Image image) {
		this(context, width, height);
		
		this.image = image;
		
		image.setCenterOfRotation(getWidth() / 2, getHeight() / 2);
	}
	
	public Sprite(GUIContext context, Image image) {
		this(context, image.getWidth(), image);
	}

	@Override
	public void render(GUIContext context, Graphics g) throws SlickException {
		if (!isVisible()) {
			return;
		}
				
		if (image != null) {
			image.draw(getX(), getY(), getWidth(), getHeight());
		}
		
		super.render(context, g);
	}
	
	public void setImage(Image image) {
		this.image = image;

		image.setCenterOfRotation(getWidth() / 2, getHeight() / 2);
	}
	
	public Image getImage(){
		return this.image;
	}
	
	@Override
	public String toString() {
		return image.getResourceReference();
	}

	public float getRotation() {
		return image.getRotation();
	}
	
	public void rotate(int delta) {
		image.setRotation((getRotation() + delta));
	}
}

File: web/src/component/Sprite.ts
import { Component } from './Component';

export class Sprite extends Component {
  private image: HTMLImageElement | null = null;
  private rotation: number = 0;

  constructor(width: number, height?: number, image?: HTMLImageElement) {
    if (image && !height) {
      const aspectRatio = image.height / image.width;
      super(width, Math.floor(width * aspectRatio));
    } else if (height !== undefined) {
      super(width, height);
    } else {
      super(width, width);
    }

    if (image) {
      this.image = image;
    }
  }

  override render(ctx: CanvasRenderingContext2D): void {
    if (!this.isVisible()) {
      return;
    }

    if (this.image !== null) {
      ctx.save();
      ctx.imageSmoothingEnabled = false;

      if (this.rotation !== 0) {
        const centerX = this.getX() + this.getWidth() / 2;
        const centerY = this.getY() + this.getHeight() / 2;
        ctx.translate(centerX, centerY);
        ctx.rotate((this.rotation * Math.PI) / 180);
        ctx.translate(-centerX, -centerY);
      }

      ctx.drawImage(this.image, this.getX(), this.getY(), this.getWidth(), this.getHeight());

      ctx.restore();
    }

    super.render(ctx);
  }

  setImage(image: HTMLImageElement): void {
    this.image = image;
  }

  getImage(): HTMLImageElement | null {
    return this.image;
  }

  getRotation(): number {
    return this.rotation;
  }

  setRotation(rotation: number): void {
    this.rotation = rotation;
  }

  rotate(delta: number): void {
    this.rotation = (this.rotation + delta) % 360;
  }

  toString(): string {
    return this.image?.src || 'Sprite[no image]';
  }
}
